Add pmi-simple server to wrexecd

grondo opened the pull request:
This PR is a quick and dirty addition of pmi-simple server to `wrexecd`. By default wrexec jobs will have `PMI_FD` etc set in environment and, if supported, simple-pmi enabled MPI and other programs will use PMI-1 wire protocol instead of a `libpmi.so` library.

To disable pmi-server, set `-o no-pmi-server` on `flux wreckrun` cmdline.

This has only had cursory testing at this point, but up for early review. It is based on top of #708

garlick commented:
Just had a quick skim and looks good, although I feel bad that the pmi-server abstraction is so weak.  Maybe we should figure out what needs to change in it to allow the write direction to use a non-blocking zio.

garlick commented:
Would a callback that libpmi-server makes for each response be slightly more helpful than the `pmi_simple_server_response()` call?

garlick commented:
One caveat that we need to address is pre-loading the KVS with the clique ranks.  Tom talks about this in issue #665.  I wanted to get the "KVS schema" precisely nailed down in RFC 13 first, so that is on me.

grondo commented:
> although I feel bad that the pmi-server abstraction is so weak. ￼Maybe we should figure out what needs to change in it to allow the write direction to use a non-blocking zio.

Hm, it wasn't so bad. I did end up using non-blocking `zio` for both the read/write sides of the client connection, though for the "write" it seemed a bit overkill (and I just realized I think I forgot to wire the client side to the reactor, but it works anyways because we never get short writes in testing).

The code for the read side is a bit overcomplicated because zio reader in line buffered mode can return multiple lines per callback so that has to be handled. With all the extra line splitting code. This could be simplified with a flag to zio that returns only one line at a time.

> Would a callback that libpmi-server makes for each response be slightly more helpful than the pmi_simple_server_response() call?

Yeah, that is what I was thinking, it could be an implementation write handler like `int client_write (void *client, const char *resp)`. Then all responses could be written as needed by pmi_server. In `flux-start` you could register `dputline` wrapper and in `wrexecd` we'd register a wrapper for `zio_write`.

garlick commented:
> I did end up using non-blocking zio for both the read/write sides of the client connection

Oh great!   Sorry I missed that it was for both directions.  Good!

> Ok, just so I'm on the same page, you mean the PMI KVS preloaded with PMI_process_mapping? This might be easiest to do from flux-wreckrun or possibly nodeid 0 wrexecd.

Yes that.  I just updated flux-framework/rfc#49 to include a description of the format FWIW.  Should there be a helper in the PMI server code to generate the string?  (Seems like it?)

grondo commented:
Another request for `pmi_simple_server` that will simplify code using it:
```C
/* Put null-terminated request with sending client reference to protocol
 * engine.  Caller should remove the trailing newline.
 * Return 0 on success, -1 on failure.
 */
int pmi_simple_server_request (struct pmi_simple_server *pmi,
                               const char *buf, void *client);
```

It would seem to make calling code simpler if `pmi_simple_server_request()` allowed/required the trailing newline in `buf`, instead of *requiring* the caller to remove it.

Since the protocol does require the newlines, this also seems to make more sense from a protocol perspective... However, I'm not sure why it was done this way, perhaps it will complicate the `flux-start` use case?

garlick commented:
I was having the same thought and I can't remember exactly why I did it that way.  Will fix.

garlick commented:
Just pushed some changes to pr #708 including the above newline changes and replacing `pmi_simple_server_response()` with a `response_send()` callback.

grondo commented:
Great thanks! I'll rebase onto your branch.

grondo commented:
Ok, rebased onto latest changes in #708 and squashed the entire thing down to 2 commits. 

Use of zio is simplified by allowing a zio to set the max # of lines we want to read per callback, and with the other changes in #708 the amount of code required to wire pmi-simple-server into wrexecd was pretty small. (Thanks, @garlick!)

grondo commented:
Rebased onto latest master.

I also added a couple experimental testing changes. 

 * enable simple MPI tests in Travis by setting `TEST_MPI=t`
 * Run MPI tests both with and without `no-pmi-server` set, to exercise both simple pmi server and libpmi.so from wreckrun.

We'll see how that goes...

grondo commented:
Might be more debugging required here. Some of the MPI tests failed in travis.

garlick commented:
Maybe failing because PMI_process_mapping is not set?

We could try setting that to the empty string for a start.  I believe that's supposed to be handled as a non-failure.

garlick commented:
Well, I can special case `PMI_process_mapping` to return an empty string in `wreck_pmi_kvs_get()` and this allows the first iteration of the tests to work.  However the second iteration with the `no-pmi-server` option are running as a group of singletons.

Just checked back and this also fails on master on my desktop with the Ubuntu mpich - which may be because that version was built without libpmi.so support.  I'm not sure about what we're building for travis.  Possibly it only supports the wire proto as well?

garlick commented:
Until wreckrun (or somebody) begins setting PMI_process_mapping in the Flux KVS, a fallback like this is probably OK:
```diff
diff --git a/src/modules/wreck/wrexecd.c b/src/modules/wreck/wrexecd.c
index 818d6ee..d5a7b60 100644
--- a/src/modules/wreck/wrexecd.c
+++ b/src/modules/wreck/wrexecd.c
@@ -1985,9 +1985,17 @@ static int wreck_pmi_kvs_get (void *arg, const char *kvsname, const char *key,
     }
 
     if (kvs_get_string (ctx->flux, kvskey, &s) < 0) {
-        wlog_err (ctx, "pmi_kvs_get: kvs_get_string: %s", strerror (errno));
-        free (kvskey);
-        return (-1);
+        /* PMI_process_mapping is required by PMI 1.1.
+         * If not set, assume the local process group is unknown.
+         * This is indicated to PMI clients by a value of "".
+         */
+        if (errno == ENOENT && !strcmp (key, "PMI_process_mapping")) {
+            s = xstrdup ("");
+        } else {
+            wlog_err (ctx, "pmi_kvs_get: kvs_get_string: %s", strerror (errno));
+            free (kvskey);
+            return (-1);
+        }
     }
 
     if (strlen (s) >= len) {
```

grondo commented:
Thanks for solving that one @garlick! Want to push that commit and I'll cherry-pick it into this branch?

It occurs to me that wreckrun or equivalent will not be able to set `PMI_process_mapping` since in the current design the "front end command" is not required to run programs. In fact it is a distributed service managed by wrexecd or the shell replacement. (e.g. when sched is enabled it will perform process mapping and request initiation of the program)

The obvious place to do this now would be in the first wrexecd participating in the launch. It will have to walk kvs directories for each rank and gather the number of tasks assigned in order to create the mapping, which will slow down the launch.

I don't think I saw any barrier around the first "get" of `PMI_process_mapping`? If not, then we'll probably have to change the "starting" job state to a barrier so that no process is launched until global PMI state has been initialized.

And this will all have to be done for every job, whether they use PMI or not. I wonder if we could trigger the creation of `PMI_process_mapping` on the first PMI call from a client to avoid wasting resources or becoming a MPI-only launcher. Or perhaps we should have a `flux mpirun`....

garlick commented:
Here's an attempt that seems to work for you to cherry pick:  79e846d40ec8a9b622694741b96edac9341bdaf5

grondo commented:
Hm would it be too much to ask to split the testing and code changes into separate commits?
Sorry :(

garlick commented:
I meant to add a tiny bit more explanation - the above just calculates a `PMI_process_mapping` value and returns it directly when it's asked for, no KVS interaction.

The value is `(vector,(0,nnodes,ntasks))` where nnodes is the total number of nodes in the job, and ntasks is the number of tasks on this node. It presumes that the value is used solely for calculating clique ranks.  Assuming a "block" mapping of ranks to tasks, this should do it I think.

I launched some MPI progams though I can't really tell if they are using shmem versus tcp to communicate.  I will poke at it a bit more and see if I tell.

My first and second attempts at this failed miserably.  The first attempt was to set three blocks: for tasks before, in, and after the current set of tasks.  That resulted in segfaults from MPI.  The second attempt was to trim the before and/or after blocks if they had no tasks.  That resulted in deadlocked MPI tasks (many of them after a while!).  I am still not too sure why this failed, and so I am a little uncomfortable with this.

(Just saw your message about splitting the commits - yes coming up! :-)

garlick commented:
New commits are
* Test -  4ad72dd8a3443f0475413f3b81ca87ab1a7105f6
* Wrexecd - aa33ad1e0c4a375806deb46b3fc8f4cac3faaaba

grondo commented:
> It presumes that the value is used solely for calculating clique ranks. Assuming a "block" mapping of ranks to tasks, this should do it I think.

A good test would be to see if this works with block distribution with unequal tasks per node. E.g. run 6 tasks across 2 "nodes" with 4 tasks on the first and 2 on the second. It would seem like the `PMI_process_mapping` value provided here would end up causing the 2nd node to think it was running ranks 2,3 when actually it is running 4,5.

I still might not understand `PMI_process_mapping` though.

grondo commented:
> New commits are

Thanks!!

garlick commented:
Oh duhh, you're right.  This is annoying.

garlick commented:
OK, back to the empty string :-(  2dcec36e934d710bb2ed58b53a1d9ef532a33592

grondo commented:
Thanks, and sorry about making you do all the work!

For testing purposes, do you think fallthrough to empty string for `PMI_process_mapping` would be better than diverting the PMI_kvs_get request? Then I can play with preset PMI_process_mapping and it will just work if it is preloaded into lwj pmikvs space.

I can go ahead and make that change if it doesn't bother you. I know it will be a bit slower...

garlick commented:
Fine with me.  I just didn't think the hit was worth it if we weren't planning to play with this for a while.
Incidentally, I just noticed that hydra seems to always set this to `(vector,(0,1,1))` which might be slightly better than the empty string as a default.

grondo commented:
>  Fine with me. I just didn't think the hit was worth it if we weren't planning to play with this for a while.

Good point. My thinking was that, if we were going for performance runs, we would probably want to figure out a way to pre-load PMI_process_mapping into pmi kvs, but I realize I don't actually know the impact of effectively disabling shmem, so maybe we can wait awhile on that?

garlick commented:
Correction FWIW: the way to alter hydra's mapping is with a host file.  Example:
```
localhost:2
jimbo:3
localhost:4
```
(those are all the same machine, but if I use the same name it tries to be smart).  Then
```
$ mpiexec.hydra -f hosts.conf -n 8
```
results in `(vector,(0,1,2),(1,1,3),(2,1,4))`

Learned something here:  the blocks don't have to add up (9 tasks defined, 8 used).

grondo commented:
> Learned something here: the blocks don't have to add up (9 tasks defined, 8 used).

This format hurts my brain. Does the oversupply of task slots in blocks only work if it is for the very last block, and only if that block is for a single node? Otherwise, how would a node determine its clique ranks if it doesn't know if nodes before it are under provisioned or exactly provisioned? This seems like perhaps a weird corner case or bug.. 

(If it did work somehow, maybe we could get away with `(vector,(0,nnodes,MAX_INT))`)

grondo commented:
Hm, I just realized that wrexecd already has to traverse all `lwj.<id>.rank.N` dirs in kvs in order to compute the global taskids of the tasks running on this node (a definite inefficiency in the hacky wreck code). Maybe we ignore the inefficiency of that for now and use that loop to compute PMI_process_mapping,  *or* another approach might be to slightly modify the format of how task to broker rank mapping is placed in the KVS (maybe using something similar to PMI_process_mapping?)

I hate to put more work into wrexec at this point, but going through an exercise of how to more efficiently setup information each shell will need to read from kvs or receive in a msg might be a good thing.

garlick commented:
Poking at it a bit, it seems like hydra defines blocks from the hosts file, fills a block with tasks before starting on the next one, includes only blocks that have any tasks allocated in them, and doesn't bother to trim the final block, if partially filled.

grondo commented:
I just pushed some *experimental* changes to wrexecd that allow generation of `pmi.PMI_process_mapping` from the rank0 wrexecd of a job via a `rexecd_init` callback in a Lua plugin. I have not tried to see if this works for MPI yet as I don't so far have a good place to test a PMI 1.1 MPI.

The changes here also adjust the number of `kvs_fence/kvs_commit` calls used by wreck startup, I think probably for the better (1 new kvs_fence was added for the "starting" state, but an unnecessary kvs_commit was removed from every other state transition). It probably won't have a noticeable effect.

garlick commented:
Hmm, tests are not acquiring the mapping
```
# ../../t/t3000-mpi-basic.t: flux session size will be 4
ok 1 - mpi hello singleton with stdio-delay-commit
2016-06-29T18:21:26.786884Z lwj.2.err[1]: pmi_kvs_get: kvs_get_string(lwj.2.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:26.788512Z lwj.2.err[3]: pmi_kvs_get: kvs_get_string(lwj.2.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:26.788886Z lwj.2.err[2]: pmi_kvs_get: kvs_get_string(lwj.2.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:26.789912Z lwj.2.err[0]: pmi_kvs_get: kvs_get_string(lwj.2.pmi.PMI_process_mapping): No such file or directory
ok 2 - mpi hello all ranks with stdio-delay-commit
2016-06-29T18:21:27.303547Z lwj.3.err[2]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.310536Z lwj.3.err[1]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.303955Z lwj.3.err[2]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.303135Z lwj.3.err[0]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.304907Z lwj.3.err[2]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.316045Z lwj.3.err[1]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.315012Z lwj.3.err[0]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.316932Z lwj.3.err[2]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.317787Z lwj.3.err[1]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.320191Z lwj.3.err[1]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.321800Z lwj.3.err[0]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.322857Z lwj.3.err[3]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.323270Z lwj.3.err[3]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.323422Z lwj.3.err[0]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.323889Z lwj.3.err[3]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
2016-06-29T18:21:27.324335Z lwj.3.err[3]: pmi_kvs_get: kvs_get_string(lwj.3.pmi.PMI_process_mapping): No such file or directory
ok 3 - mpi hello oversubscribed with stdio-delay-commit
```

Also is the lua plugin not storing the "vector" type?  That's required I believe, e.g.
```
(vector,(a,b,c),(d,e,f),...)
```

grondo commented:
> Also is the lua plugin not storing the "vector" type? That's required I believe, e.g.

It is, and I'm not sure why the mapping isn't in the kvs. It is when I test locally. Hm, I wonder if I forgot to add the new plugin to a Makefile.am somewhere so it actually isn't being run in testing. oops.

I don't think the mpi tests are going to work here because MPI is a big unknown.
What we should probably do is write a small standalone test that verifies PMI wire protocol works. Then we can run against that in wreck tests as well as standalone pmi-server tests.

Also, if we plan to continue to support `libpmi.so` based bootstrap, we probably need a similar test that verifies that interface (explicitly does not use PMI_FD).s

grondo commented:
For now I'll back out the changes to `t3000-mpi-basic.t`, but we can still run in Travis with `TEST_MPI` set. This should be an ok sanity check for pmi-simple-server embedded in wrexecd. Then this might be ready for merge, but I'd feel better if @garlick could verify that `PMI_process_mapping` is actually getting used.

garlick commented:
OK let me see if I can validate that by stracing the mpi executable.

grondo commented:
Ok, I kept the t3000-mpi-basic.t change, but removed `"stdio-delay-commit,no-pmi-server"` from the list of `$OPTS` to iterate (so really we don't iterate anymore). This should at least exercise MPI startup under Flux, and the code is there to run with other options if necessary in the future.

I also added one test to t2002-pmi.t to verify that PMI_process_mapping at least exists in kvs space of the job, and that a simple case has the expected value.

garlick commented:
Looking through traces generated by these two cases:
```
$ flux start flux wreckrun -N1 -n2 strace -ff -o /tmp/trace_shm mpi_hello
$ flux start -s2 flux wreckrun -N2 -n2 strace -ff -o /tmp/trace_tcp mpi_hello
```
Sorry for the verbosity but for completeness, here's a trace excerpt from one rank of each test covering the whole pmi conversation.  It looks to me like the first mmaps a shmem area while the second does a lot of socket stuff, as expected.

**shmem**
```
write(24, "cmd=init pmi_version=1 pmi_subve"..., 40) = 40
read(24, "cmd=response_to_init pmi_version"..., 1023) = 57
write(24, "cmd=get_maxes\n", 14)        = 14
read(24, "cmd=maxes kvsname_max=64 keylen_"..., 1023) = 54
write(24, "cmd=get_appnum\n", 15)       = 15
read(24, "cmd=appnum appnum=1\n", 1023) = 20
write(24, "cmd=get_my_kvsname\n", 19)   = 19
read(24, "cmd=my_kvsname kvsname=lwj.1.pmi"..., 1023) = 33
write(24, "cmd=get_my_kvsname\n", 19)   = 19
read(24, "cmd=my_kvsname kvsname=lwj.1.pmi"..., 1023) = 33
write(24, "cmd=get kvsname=lwj.1.pmi key=PM"..., 50) = 50
read(24, "cmd=get_result rc=0 msg=success "..., 1023) = 55
uname({sysname="Linux", nodename="jimbo", ...}) = 0
write(24, "cmd=barrier_in\n", 15)       = 15
read(24, "cmd=barrier_out\n", 1023)     = 16
write(24, "cmd=get kvsname=lwj.1.pmi key=sh"..., 48) = 48
read(24, "cmd=get_result rc=0 msg=success "..., 1023) = 68
open("/dev/shm/mpich_shar_tmpYbGKbb", O_RDWR) = 3
close(3)                                = 0
open("/dev/shm/mpich_shar_tmpYbGKbb", O_RDWR) = 3
mmap(NULL, 8522576, PROT_READ|PROT_WRITE, MAP_SHARED, 3, 0) = 0x7fd991f41000
close(3)                                = 0
socket(PF_INET, SOCK_STREAM, IPPROTO_TCP) = 3
setsockopt(3, SOL_TCP, TCP_NODELAY, [1], 4) = 0
getsockopt(3, SOL_TCP, TCP_NODELAY, [1], [4]) = 0
setsockopt(3, SOL_SOCKET, SO_RCVBUF, [131072], 4) = 0
getsockopt(3, SOL_SOCKET, SO_RCVBUF, [262144], [4]) = 0
setsockopt(3, SOL_SOCKET, SO_SNDBUF, [262144], 4) = 0
getsockopt(3, SOL_SOCKET, SO_SNDBUF, [425984], [4]) = 0
fcntl(3, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(3, F_SETFL, O_RDWR)               = 0
fcntl(3, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(3, F_SETFL, O_RDWR|O_NONBLOCK)    = 0
bind(3, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("0.0.0.0")}, 16) = 0
listen(3, 128)                          = 0
uname({sysname="Linux", nodename="jimbo", ...}) = 0
socket(PF_INET, SOCK_DGRAM, IPPROTO_IP) = 4
ioctl(4, SIOCGIFCONF, {10 * sizeof(struct ifreq) => 2 * sizeof(struct ifreq), [{ifr_name="lo", ifr_addr={AF_INET, inet_addr("127.0.0.1")}}, {ifr_name="enp6s0", ifr_addr={AF_INET, inet_addr("192.168.1.136")}}]}) = 0
ioctl(4, SIOCGIFCONF, {20 * sizeof(struct ifreq) => 2 * sizeof(struct ifreq), [{ifr_name="lo", ifr_addr={AF_INET, inet_addr("127.0.0.1")}}, {ifr_name="enp6s0", ifr_addr={AF_INET, inet_addr("192.168.1.136")}}]}) = 0
close(4)                                = 0
getsockname(3, {sa_family=AF_INET, sin_port=htons(51668), sin_addr=inet_addr("0.0.0.0")}, [16]) = 0
rt_sigaction(SIGPIPE, {SIG_IGN, [PIPE], SA_RESTORER|SA_RESTART, 0x7fd9931e04a0}, {SIG_DFL, [], 0}, 8) = 0
write(24, "cmd=put kvsname=lwj.1.pmi key=P1"..., 103) = 103
read(24, "cmd=put_result rc=0 msg=success\n", 1023) = 32
write(24, "cmd=barrier_in\n", 15)       = 15
read(24, "cmd=barrier_out\n", 1023)     = 16
rt_sigaction(SIGUSR1, {0x7fd9936c1ed0, [USR1], SA_RESTORER|SA_RESTART, 0x7fd9931e04a0}, {SIG_DFL, [], 0}, 8) = 0
mmap(NULL, 2756608, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fd991ca0000
mmap(NULL, 1314816, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fd993abe000
poll([{fd=3, events=POLLIN}], 1, 0)     = 0 (Timeout)
poll([{fd=3, events=POLLIN}], 1, 0)     = 0 (Timeout)
close(3)                                = 0
munmap(0x7fd991f41000, 8522576)         = 0
write(24, "cmd=finalize\n", 13)         = 13
read(24, "cmd=finalize_ack\n", 1023)    = 17
```

**non-shmem**
```
write(16, "cmd=init pmi_version=1 pmi_subve"..., 40) = 40
read(16, "cmd=response_to_init pmi_version"..., 1023) = 57
write(16, "cmd=get_maxes\n", 14)        = 14
read(16, "cmd=maxes kvsname_max=64 keylen_"..., 1023) = 54
write(16, "cmd=get_appnum\n", 15)       = 15
read(16, "cmd=appnum appnum=1\n", 1023) = 20
write(16, "cmd=get_my_kvsname\n", 19)   = 19
read(16, "cmd=my_kvsname kvsname=lwj.1.pmi"..., 1023) = 33
write(16, "cmd=get_my_kvsname\n", 19)   = 19
read(16, "cmd=my_kvsname kvsname=lwj.1.pmi"..., 1023) = 33
write(16, "cmd=get kvsname=lwj.1.pmi key=PM"..., 50) = 50
read(16, "cmd=get_result rc=0 msg=success "..., 1023) = 55
uname({sysname="Linux", nodename="jimbo", ...}) = 0
mmap(NULL, 4198400, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f239c288000
write(16, "cmd=barrier_in\n", 15)       = 15
read(16, "cmd=barrier_out\n", 1023)     = 16
socket(PF_INET, SOCK_STREAM, IPPROTO_TCP) = 3
setsockopt(3, SOL_TCP, TCP_NODELAY, [1], 4) = 0
getsockopt(3, SOL_TCP, TCP_NODELAY, [1], [4]) = 0
setsockopt(3, SOL_SOCKET, SO_RCVBUF, [131072], 4) = 0
getsockopt(3, SOL_SOCKET, SO_RCVBUF, [262144], [4]) = 0
setsockopt(3, SOL_SOCKET, SO_SNDBUF, [262144], 4) = 0
getsockopt(3, SOL_SOCKET, SO_SNDBUF, [425984], [4]) = 0
fcntl(3, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(3, F_SETFL, O_RDWR)               = 0
fcntl(3, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(3, F_SETFL, O_RDWR|O_NONBLOCK)    = 0
bind(3, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("0.0.0.0")}, 16) = 0
listen(3, 128)                          = 0
uname({sysname="Linux", nodename="jimbo", ...}) = 0
socket(PF_INET, SOCK_DGRAM, IPPROTO_IP) = 4
ioctl(4, SIOCGIFCONF, {10 * sizeof(struct ifreq) => 2 * sizeof(struct ifreq), [{ifr_name="lo", ifr_addr={AF_INET, inet_addr("127.0.0.1")}}, {ifr_name="enp6s0", ifr_addr={AF_INET, inet_addr("192.168.1.136")}}]}) = 0
ioctl(4, SIOCGIFCONF, {20 * sizeof(struct ifreq) => 2 * sizeof(struct ifreq), [{ifr_name="lo", ifr_addr={AF_INET, inet_addr("127.0.0.1")}}, {ifr_name="enp6s0", ifr_addr={AF_INET, inet_addr("192.168.1.136")}}]}) = 0
close(4)                                = 0
getsockname(3, {sa_family=AF_INET, sin_port=htons(50928), sin_addr=inet_addr("0.0.0.0")}, [16]) = 0
rt_sigaction(SIGPIPE, {SIG_IGN, [PIPE], SA_RESTORER|SA_RESTART, 0x7f239d1074a0}, {SIG_DFL, [], 0}, 8) = 0
write(16, "cmd=put kvsname=lwj.1.pmi key=P0"..., 103) = 103
read(16, "cmd=put_result rc=0 msg=success\n", 1023) = 32
write(16, "cmd=barrier_in\n", 15)       = 15
read(16, "cmd=barrier_out\n", 1023)     = 16
rt_sigaction(SIGUSR1, {0x7f239d5e8ed0, [USR1], SA_RESTORER|SA_RESTART, 0x7f239d1074a0}, {SIG_DFL, [], 0}, 8) = 0
mmap(NULL, 2756608, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f239bfe7000
mmap(NULL, 1314816, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f239d9e5000
write(1, "0: completed MPI_Init in 0.008s."..., 52) = 52
write(16, "cmd=get kvsname=lwj.1.pmi key=P1"..., 46) = 46
read(16, "cmd=get_result rc=0 msg=success "..., 1023) = 89
socket(PF_INET, SOCK_STREAM, IPPROTO_IP) = 4
setsockopt(4, SOL_TCP, TCP_NODELAY, [1], 4) = 0
getsockopt(4, SOL_TCP, TCP_NODELAY, [1], [4]) = 0
setsockopt(4, SOL_SOCKET, SO_RCVBUF, [131072], 4) = 0
getsockopt(4, SOL_SOCKET, SO_RCVBUF, [262144], [4]) = 0
setsockopt(4, SOL_SOCKET, SO_SNDBUF, [262144], 4) = 0
getsockopt(4, SOL_SOCKET, SO_SNDBUF, [425984], [4]) = 0
fcntl(4, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(4, F_SETFL, O_RDWR)               = 0
fcntl(4, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(4, F_SETFL, O_RDWR|O_NONBLOCK)    = 0
connect(4, {sa_family=AF_INET, sin_port=htons(38765), sin_addr=inet_addr("192.168.1.136")}, 16) = -1 EINPROGRESS (Operation now in progress)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN|POLLOUT}], 2, 0) = 1 ([{fd=4, revents=POLLOUT}])
getsockopt(4, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN|POLLOUT}], 2, 0) = 1 ([{fd=4, revents=POLLOUT}])
writev(4, [{"\0\0\0\0\0\0\0\0\4\0\0\0\0\0\0\0", 16}, {"\0\0\0\0", 4}], 2) = 20
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}], 2, 0) = 1 ([{fd=3, revents=POLLIN}])
accept(3, {sa_family=AF_INET, sin_port=htons(57406), sin_addr=inet_addr("192.168.1.136")}, [16]) = 5
setsockopt(5, SOL_TCP, TCP_NODELAY, [1], 4) = 0
getsockopt(5, SOL_TCP, TCP_NODELAY, [1], [4]) = 0
setsockopt(5, SOL_SOCKET, SO_RCVBUF, [131072], 4) = 0
getsockopt(5, SOL_SOCKET, SO_RCVBUF, [262144], [4]) = 0
setsockopt(5, SOL_SOCKET, SO_SNDBUF, [262144], 4) = 0
getsockopt(5, SOL_SOCKET, SO_SNDBUF, [425984], [4]) = 0
fcntl(5, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(5, F_SETFL, O_RDWR)               = 0
fcntl(5, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(5, F_SETFL, O_RDWR|O_NONBLOCK)    = 0
accept(3, 0x7fff993bfbb0, 0x7fff993bfba8) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN}], 3, 0) = 1 ([{fd=5, revents=POLLIN}])
getsockopt(5, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
read(5, "\0\0\0\0\0\0\0\0\4\0\0\0\0\0\0\0", 16) = 16
readv(5, [{"\1\0\0\0", 4}], 1)          = 4
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN|POLLOUT}], 3, 0) = 1 ([{fd=5, revents=POLLOUT}])
getsockopt(5, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN|POLLOUT}], 3, 0) = 1 ([{fd=5, revents=POLLOUT}])
getsockopt(5, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=5, events=POLLIN|POLLOUT}], 3, 0) = 2 ([{fd=4, revents=POLLIN}, {fd=5, revents=POLLOUT}])
read(4, "\2\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 16
close(4)                                = 0
getsockopt(5, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
write(5, "\1\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 16
writev(5, [{"\0\0\0\0\0\0\0\0\1\0\0\0\0\0\1\0\1\0\0\254\0\0\0\0\0\0\0\0\0\0\0\0"..., 48}], 1) = 48
poll([{fd=3, events=POLLIN}, {fd=-1}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=-1}, {fd=5, events=POLLIN}], 3, 0) = 0 (Timeout)
poll([{fd=3, events=POLLIN}, {fd=-1}, {fd=5, events=POLLIN}], 3, 0) = 1 ([{fd=5, revents=POLLIN}])
recvfrom(5, "\0\0\0\0\0\0\0\0\1\0\0\0\1\0\1\0\1\0\0\254\0\0\0\0\0\0\0\0\0\0\0\0"..., 1024, 0, NULL, NULL) = 48
write(1, "0: completed first barrier in 0."..., 37) = 37
writev(5, [{"#\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 48}, {NULL, 0}], 2) = 48
poll([{fd=3, events=POLLIN}, {fd=-1}, {fd=5, events=POLLIN}], 3, 0) = 1 ([{fd=5, revents=POLLIN}])
recvfrom(5, "#\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 1024, 0, NULL, NULL) = 48
writev(5, [{"#\0\0\0\1\0\0\0h\321\264\235#\177\0\0x\375;\231\377\177\0\0O\t\223\235#\177\0\0"..., 48}, {NULL, 0}], 2) = 48
poll([{fd=3, events=POLLIN}, {fd=-1}, {fd=5, events=POLLIN}], 3, 0) = 1 ([{fd=5, revents=POLLIN}])
recvfrom(5, "#\0\0\0\1\0\0\0h\221\351\336\336\177\0\0\10$\f\252\375\177\0\0O\311\307\336\336\177\0\0"..., 1024, 0, NULL, NULL) = 48
close(5)                                = 0
close(3)                                = 0
munmap(0x7f239c288000, 4198400)         = 0
write(16, "cmd=finalize\n", 13)         = 13
read(16, "cmd=finalize_ack\n", 1023)    = 17
```

garlick commented:
Nice work, merging.
